### PR TITLE
Add optional dependency on total-perspective-vortex

### PR DESCRIPTION
**Pull request by nuwang**
This PR adds [total-perspective-vortex](https://github.com/galaxyproject/total-perspective-vortex) (TPV) as an optional dependency.

TPV provides an installable set of dynamic rules that can route entities (Tools, Users, Roles) to appropriate job destinations based on a configurable yaml file. TPV also provides pluggable meta-scheduling capabilities, and a centralizable database of rules that can be shared by multiple Galaxy instances. It is meant to be a successor to Galaxy's dynamic tool destinations.

TPV is in production in Galaxy-Australia, and we've seen a significant increase in throughput (2x) due to better scheduling and utilization of Pulsar nodes.

More information: [http://total-perspective-vortex.readthedocs.org](http://total-perspective-vortex.readthedocs.org/en/latest/?badge=latest)

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

**Comment by bernt-matthias** [comment]
This sounds very interesting. Does tpv also cover job resubmission?

**Comment by nuwang** [comment]
@bernt-matthias I can't see any reason why job resubmission wouldn't work, but I haven't tested this myself. Will get back to you on it.

File: test/unit/app/jobs/job_conf.sample_advanced.yml [hunk 1]
```diff
@@ -749,6 +749,17 @@ execution:
...
+      runner: dynamic
+      # Demonstrates how to use total-perspective-vortex, which enables dynamic destination
+      # selection for entities (Tools, Users, Groups) based on a configurable yaml file.
+      # See: https://total-perspective-vortex.readthedocs.io/ for documentation.
+      type: python
+      function: map_tool_to_destination
+      rules_module: vortex.rules
+      vortex_config_files:
+        - https://github.com/galaxyproject/total-perspective-vortex/blob/bf8ec46c672ed5a7ba2471d9b514cef15995bf3f/tests/fixtures/mapping-rules.yml
+        - config/vortex_rules_local.yml
```
Comment (afgane) [inline test/unit/app/jobs/job_conf.sample_advanced.yml]: The first item should be removed (or commented out) and a corresponding file for the second line should be added that can be readily used with a local runner.

With the following `job_conf.yml`:
```
runners:
  local:
    load: galaxy.jobs.runners.local:LocalJobRunner
    workers: 4

execution:
  default: vortex_dispatcher
  environments:
    vortex_dispatcher:
      runner: dynamic
      type: python
      function: map_tool_to_destination
      rules_module: vortex.rules
      vortex_config_files:
        - config/mapping-rules.yml
    local:
      runner: local
```

A seemingly a minimal mapping file is as follows, jobs don't run (`'No destinations are available to fulfill request: default'`) so I'm guessing others will run into the same challenge of trying to understand how all fits together before seeing anything run:
```
global:
  default_inherits: default

tools:
  default:
    cores: 2
    mem: cores * 2

destinations:
  local:
    cores: 1
    mem: 2
    scheduling:
      prefer:
        - general

```

File: lib/galaxy/dependencies/conditional-requirements.txt [hunk 2]
```diff
@@ -12,6 +12,7 @@ python-irodsclient==1.0.0
 python-ldap==3.4.0
 python-pam
 galaxycloudrunner
+total-perspective-vortex
```
Comment (afgane) [inline lib/galaxy/dependencies/conditional-requirements.txt:15]: I don't know if this is just my local setup (which was a fresh clone) or something unrelated to this PR, but this package did not automatically install following relevant configuration updates.

**Review by afgane — CHANGES_REQUESTED** [review]

File: lib/galaxy/dependencies/conditional-requirements.txt (hunk 2 above)
Comment (nuwang) [reply lib/galaxy/dependencies/conditional-requirements.txt:15]: I tested this and it did get installed. I think I added `job_config_file: config/job_conf.yml` in `galaxy.yml` so maybe a local config issue?

File: test/unit/app/jobs/job_conf.sample_advanced.yml (hunk 1 above)
Comment (nuwang) [reply test/unit/app/jobs/job_conf.sample_advanced.yml]: Have added a dedicated sample file that should work for a local destination out of the box.

**Review by afgane — APPROVED** [review]
Tested locally and works nicely. Test failure seems unrelated.

**Comment by nuwang** [comment]
@bernt-matthias We discussed this briefly, and there could be some issues with resubmission. It's on the backlog of tasks to work on and the plan is to add support for resubmissions should it not be working already.

**Comment by nuwang** [comment]
@bernt-matthias We've added in-built support for resubmits, as well as integration tests for it: https://github.com/galaxyproject/total-perspective-vortex/pull/22

**Pull request merged**